pkg/scrape/discovery/consul/consul.go

```diff
@@ -0,0 +1,583 @@
+	finalizer        func()
+	logger           logrus.FieldLogger
+}
+
+// NewDiscovery returns a new Discovery for the given config.
+func NewDiscovery(conf *SDConfig, logger logrus.FieldLogger) (*Discovery, error) {
+	wrapper, err := config.NewClientFromConfig(conf.HTTPClientConfig, "consul_sd", config.WithIdleConnTimeout(2*watchTimeout))
+	if err != nil {
+		return nil, err
+	}
+	wrapper.Timeout = watchTimeout + 15*time.Second
+
+	clientConf := &consul.Config{
+		Address:    conf.Server,
+		Scheme:     conf.Scheme,
+		Datacenter: conf.Datacenter,
+		Namespace:  conf.Namespace,
+		Partition:  conf.Partition,
+		Token:      string(conf.Token),
+		HttpClient: wrapper,
+	}
+	client, err := consul.NewClient(clientConf)
+	if err != nil {
+		return nil, err
+	}
+	cd := &Discovery{
+		client:           client,
+		tagSeparator:     conf.TagSeparator,
+		watchedServices:  conf.Services,
+		watchedTags:      conf.ServiceTags,
+		watchedNodeMeta:  conf.NodeMeta,
+		allowStale:       conf.AllowStale,
+		refreshInterval:  time.Duration(conf.RefreshInterval),
+		clientDatacenter: conf.Datacenter,
+		clientNamespace:  conf.Namespace,
+		clientPartition:  conf.Partition,
+		finalizer:        wrapper.CloseIdleConnections,
+		logger:           logger,
+	}
+	return cd, nil
+}
+
+// shouldWatch returns whether the service of the given name should be watched.
+func (d *Discovery) shouldWatch(name string, tags []string) bool {
+	return d.shouldWatchFromName(name) && d.shouldWatchFromTags(tags)
+}
+
+// shouldWatch returns whether the service of the given name should be watched based on its name.
+func (d *Discovery) shouldWatchFromName(name string) bool {
+	// If there's no fixed set of watched services, we watch everything.
+	if len(d.watchedServices) == 0 {
+		return true
+	}
+
+	for _, sn := range d.watchedServices {
+		if sn == name {
+			return true
+		}
+	}
+	return false
+}
+
+// shouldWatch returns whether the service of the given name should be watched based on its tags.
+// This gets called when the user doesn't specify a list of services in order to avoid watching
+// *all* services. Details in https://github.com/prometheus/prometheus/pull/3814
+func (d *Discovery) shouldWatchFromTags(tags []string) bool {
+	// If there's no fixed set of watched tags, we watch everything.
+	if len(d.watchedTags) == 0 {
+		return true
+	}
+
+tagOuter:
+	for _, wtag := range d.watchedTags {
+		for _, tag := range tags {
+			if wtag == tag {
+				continue tagOuter
+			}
+		}
+		return false
+	}
+	return true
+}
+
+// Get the local datacenter if not specified.
+func (d *Discovery) getDatacenter() error {
+	// If the datacenter was not set from clientConf, let's get it from the local Consul agent
+	// (Consul default is to use local node's datacenter if one isn't given for a query).
+	if d.clientDatacenter != "" {
+		return nil
+	}
+
+	info, err := d.client.Agent().Self()
+	if err != nil {
+		d.logger.WithError(err).Error("Error retrieving datacenter name")
```

Comment:
```suggestion
		d.logger.WithError(err).Error("error retrieving datacenter name")
```
Sorry for nitpicking but could you please also make log messages begin with a lowercase letter – just for consistency with other log messages in Pyroscope

Reply:
Fixed~